binr/r2r/load.c

```diff
@@ -0,0 +1,160 @@
+/* radare - LGPL - Copyright 2020 - thestr4ng3r */
+
+#include "r2r.h"
+
+#define LINEFMT "%s, line %"PFMT64u": "
+
+R_API R2RCmdTest *r2r_cmd_test_new() {
+	return R_NEW0 (R2RCmdTest);
+}
+
+R_API void r2r_cmd_test_free(R2RCmdTest *test) {
+	if (!test) {
+		return;
+	}
+#define DO_KEY_STR(key, field) free (test->field.value);
+	R2R_CMD_TEST_FOREACH_RECORD(DO_KEY_STR, R2R_CMD_TEST_FOREACH_RECORD_NOP)
+#undef DO_KEY_STR
+	free (test);
+}
+
+// read the (possibly multiline) string value of some key in the file
+// e.g. for
+//
+// 0    CMDS=<<EOF
+// 1    Hello
+// 2    World
+// 3    EOF
+// 4    ...
+//
+// if f is at the beginning of line 1,
+// read_string_val(f, "<<EOF\0")
+// will return "Hello\nWorld\n" with f being at the beginning of line 4 afterwards.
+static char *read_string_val(FILE *f, const char *val, ut64 *linenum) {
+	if (val[0] == '\'') {
+		eprintf ("Error: Invalid string syntax, use <<EOF instead of '...'\n");
+		return NULL;
+	}
+	if (val[0] == '<' && val[1] == '<') {
+		// <<EOF syntax
+		const char *endtoken = val + 2;
+		if (!*endtoken) {
+			eprintf ("Error: Missing opening end token after <<\n");
+			return NULL;
+		}
+		RStrBuf *buf = r_strbuf_new ("");
+		char *line = NULL;
+		size_t linesz = 0;
+		ssize_t len;
+		while ((len = getline (&line, &linesz, f)) >= 0) {
+			(*linenum)++;
+			char *end = strstr (line, endtoken);
+			if (end) {
+				*end = '\0';
+			}
+			r_strbuf_append (buf, line);
+			if (end) {
+				break;
+			}
+		}
+		free (line);
+		if (len < 0) {
+			eprintf ("Error: Missing closing end token %s\n", endtoken);
+			r_strbuf_free (buf);
+			return NULL;
+		}
+		return r_strbuf_drain (buf);
+	}
+	return strdup (val);
+}
+
+R_API RPVector *r2r_load_cmd_test_file(const char *file) {
+	FILE *f = r_sandbox_fopen (file, "r");
+	if (!f) {
+		eprintf ("Failed to open file \"%s\"\n", file);
+		return NULL;
+	}
+
+	RPVector *ret = r_pvector_new ((RPVectorFree)r2r_cmd_test_free);
+	if (!ret) {
+		return NULL;
+	}
+	R2RCmdTest *test = r2r_cmd_test_new ();
+	if (!test) {
+		r_pvector_free (ret);
+		return NULL;
+	}
+
+	ut64 linenum = 0;
+	char *line = NULL;
+	size_t linesz = 0;
+	ssize_t len;
+	while ((len = getline (&line, &linesz, f)) >= 0) {
+		linenum++;
+		if (!len) {
+			continue;
+		}
+		if (line[len - 1] == '\n') {
+			if (len == 1) {
+				continue;
+			}
+			line[len - 1] = '\0';
+		}
+		char *val = strchr (line, '=');
+		if (val) {
+			*val = '\0';
+			val++;
+		}
+
+		// RUN is the only cmd without value
+		if (strcmp (line, "RUN") == 0) {
+			r_pvector_push (ret, test);
+			test = r2r_cmd_test_new ();
+			if (!test) {
+				goto beach;
+			}
+			continue;
+		}
+
+#define DO_KEY_STR(key, field) \
```

Comment:
multiline macros are crap. get rid of this mess

Reply:
The alternative is only code duplication and replicating the same list of fields across multiple code parts while `R2R_CMD_TEST_FOREACH_RECORD()` allows them to be specified in a single place. So I prefer it this way.